src/vmm/src/builder.rs

```diff
@@ -470,27 +491,96 @@ pub fn build_microvm_from_snapshot(
     debug!("event_start: build microvm from snapshot");
+
+    let secret_free = vm_resources.machine_config.secret_free;
+
+    let mut kvm_capabilities = microvm_state.kvm_state.kvm_cap_modifiers.clone();
+    if secret_free {
+        kvm_capabilities.push(KvmCapability::Add(KVM_CAP_USERFAULT));
+    }
+
     let (mut vmm, mut vcpus) = create_vmm_and_vcpus(
         instance_info,
         event_manager,
         vm_resources.machine_config.vcpu_count,
-        microvm_state.kvm_state.kvm_cap_modifiers.clone(),
-        false,
+        kvm_capabilities,
+        secret_free,
     )
     .map_err(StartMicrovmError::Internal)?;
 
+    let guest_memfd = match secret_free {
+        true => Some(
+            vmm.vm
+                .create_guest_memfd(vm_resources.memory_size(), KVM_GMEM_NO_DIRECT_MAP)
+                .map_err(VmmError::Vm)?,
+        ),
+        false => None,
+    };
+
+    let userfault_bitmap_memfd = if secret_free {
+        let bitmap_size = vm_resources.memory_size() / host_page_size();
+        let bitmap_file = create_memfd(bitmap_size as u64, None)?;
+
+        // Set all bits so a fault on any page will cause a VM exit
+        let all_set = vec![0xffu8; bitmap_size];
+        bitmap_file
+            .as_file()
+            .write_all(&all_set)
+            .expect("failed to write");
+
+        Some(bitmap_file.into_file())
+    } else {
+        None
+    };
+
+    let mem_backend_path = &params.mem_backend.backend_path;
+    let mem_state = &microvm_state.vm_state.memory;
+    let track_dirty_pages = params.enable_diff_snapshots;
+
+    let (guest_memory, uffd, socket) = match params.mem_backend.backend_type {
+        MemBackendType::File => {
+            if vm_resources.machine_config.huge_pages.is_hugetlbfs() {
+                return Err(BuildMicrovmFromSnapshotErrorGuestMemoryError::File(
+                    GuestMemoryFromFileError::HugetlbfsSnapshot,
+                )
+                .into());
+            }
+            (
+                guest_memory_from_file(mem_backend_path, mem_state, track_dirty_pages)
+                    .map_err(BuildMicrovmFromSnapshotErrorGuestMemoryError::File)?,
+                None,
+                None,
+            )
+        }
+        MemBackendType::Uffd => guest_memory_from_uffd(
+            mem_backend_path,
+            mem_state,
+            track_dirty_pages,
+            vm_resources.machine_config.huge_pages,
+            guest_memfd,
+            userfault_bitmap_memfd.as_ref(),
+        )
+        .map_err(BuildMicrovmFromSnapshotErrorGuestMemoryError::Uffd)?,
+    };
+
     vmm.vm
-        .register_memory_regions(guest_memory, None)
+        .register_memory_regions(guest_memory, userfault_bitmap_memfd.as_ref())
         .map_err(VmmError::Vm)
         .map_err(StartMicrovmError::Internal)?;
     vmm.uffd = uffd;
+    vmm.uffd_socket = socket;
+
+    #[cfg(target_arch = "x86_64")]
+    vmm.vm.set_memory_private().map_err(VmmError::Vm)?;
```

Comment:
can you move this to its own commit so its easier to drop it once its not needed anymore?